File: go/utils/security.go
package utils

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// ResolveSafePath resolves a path and ensures it stays within the working directory.
// Returns the resolved path if valid, or an error if the path escapes the working directory.
func ResolveSafePath(filePath, workDir string) (string, error) {
	work, err := filepath.Abs(workDir)
	if err != nil {
		return "", fmt.Errorf("failed to resolve working directory: %w", err)
	}

	// Handle absolute paths
	if filepath.IsAbs(filePath) {
		// Check if the absolute path is within the working directory
		rel, err := filepath.Rel(work, filePath)
		if err != nil {
			return "", fmt.Errorf("path '%s' is outside working directory '%s'", filePath, workDir)
		}
		if strings.HasPrefix(rel, "..") {
			return "", fmt.Errorf("path '%s' escapes working directory '%s'", filePath, workDir)
		}
		return filePath, nil
	}

	// Join with working directory and clean the path
	target := filepath.Join(work, filePath)
	resolved, err := filepath.Abs(target)
	if err != nil {
		return "", fmt.Errorf("failed to resolve path: %w", err)
	}

	// Check if resolved path is within working directory
	rel, err := filepath.Rel(work, resolved)
	if err != nil {
		return "", fmt.Errorf("path '%s' is outside working directory '%s'", filePath, workDir)
	}
	if strings.HasPrefix(rel, "..") {
		return "", fmt.Errorf("path '%s' escapes working directory '%s'", filePath, workDir)
	}

	return resolved, nil
}

// CheckSecurityPolicy checks if text passes the whitelist/blacklist policy.
// Whitelist is checked first - if matched, allow.
// Then blacklist is checked - if matched, deny.
// Returns (allowed, reason).
func CheckSecurityPolicy(text string, whitelist, blacklist []Pattern, defaultAllow bool) (bool, string) {
	// Check whitelist first - if matched, allow
	for _, pattern := range whitelist {
		if pattern.Matches(text) {
			return true, fmt.Sprintf("Matched whitelist pattern: %s", pattern)
		}
	}

	// Check blacklist - if matched, deny
	for _, pattern := range blacklist {
		if pattern.Matches(text) {
			return false, fmt.Sprintf("Matched blacklist pattern: %s", pattern)
		}
	}

	// Default action
	if defaultAllow {
		return true, "Default allow"
	}
	return false, "Default deny"
}

// Pattern represents a whitelist or blacklist pattern.
type Pattern struct {
	Pattern string `json:"pattern"`
	Type    string `json:"type"` // "wildcard" or "regex"
}

// Matches checks if the given text matches this pattern.
func (p Pattern) Matches(text string) bool {
	switch p.Type {
	case "regex":
		return matchesRegex(p.Pattern, text)
	case "wildcard", "":
		// Use a simple wildcard matcher that supports * and ?
		return wildcardMatch(p.Pattern, text)
	default:
		return false
	}
}

// wildcardMatch implements simple wildcard matching with * and ? support.
func wildcardMatch(pattern, text string) bool {
	// Empty pattern matches empty text
	if pattern == "" {
		return text == ""
	}

	// If pattern has no wildcards, do substring match
	if !strings.Contains(pattern, "*") && !strings.Contains(pattern, "?") {
		return strings.Contains(text, pattern)
	}

	// Use filepath.Match for standard wildcard matching
	matched, err := filepath.Match(pattern, text)
	if err == nil && matched {
		return true
	}

	// Also check if pattern matches as a prefix/suffix
	if strings.HasSuffix(pattern, "*") {
		prefix := pattern[:len(pattern)-1]
		if strings.HasPrefix(text, prefix) {
			return true
		}
	}
	if strings.HasPrefix(pattern, "*") {
		suffix := pattern[1:]
		if strings.HasSuffix(text, suffix) {
			return true
		}
	}
	if strings.HasPrefix(pattern, "*") && strings.HasSuffix(pattern, "*") {
		mid := pattern[1 : len(pattern)-1]
		if strings.Contains(text, mid) {
			return true
		}
	}

	// Fallback to substring match
	return strings.Contains(text, pattern)
}

// String returns a string representation of the pattern.
func (p Pattern) String() string {
	if p.Type != "" {
		return fmt.Sprintf("%s(%s)", p.Type, p.Pattern)
	}
	return p.Pattern
}

// matchesRegex checks if text matches the given regex pattern.
func matchesRegex(pattern, text string) bool {
	// Simple regex matching using strings for basic patterns
	// For full regex support, this would use regexp.Compile
	// But to avoid import cycles and keep it simple, we do basic matching
	if strings.HasPrefix(pattern, "^") && strings.HasSuffix(pattern, "$") {
		inner := pattern[1 : len(pattern)-1]
		return text == inner
	}
	if strings.HasPrefix(pattern, "^") {
		return strings.HasPrefix(text, pattern[1:])
	}
	if strings.HasSuffix(pattern, "$") {
		return strings.HasSuffix(text, pattern[:len(pattern)-1])
	}
	return strings.Contains(text, pattern)
}

// EnsureDir ensures that the directory for the given path exists.
func EnsureDir(path string) error {
	dir := filepath.Dir(path)
	if dir == "" || dir == "." {
		return nil
	}
	return os.MkdirAll(dir, 0755)
}

// WriteFile writes data to a file, creating it if necessary.
func WriteFile(path string, data []byte) error {
	return os.WriteFile(path, data, 0644)
}

File: rust/src/utils/security.rs
//! Security utilities for path validation and command filtering.

use std::path::{Path, PathBuf};

use crate::config::Pattern;

/// Resolve a path and ensure it stays within the working directory.
///
/// Returns the resolved path if valid.
/// Returns an error if the path escapes the working directory.
pub fn resolve_safe_path(file_path: impl AsRef<Path>, work_dir: impl AsRef<Path>) -> Result<PathBuf, String> {
    let work = work_dir.as_ref().canonicalize()
        .map_err(|e| format!("Failed to canonicalize work_dir: {}", e))?;
    let target = if file_path.as_ref().is_absolute() {
        file_path.as_ref().to_path_buf()
    } else {
        work.join(file_path.as_ref())
    };
    let resolved = target.canonicalize()
        .or_else(|_| {
            // If the path doesn't exist yet (e.g., for write operations),
            // resolve the parent and append the file name
            if let Some(parent) = target.parent() {
                let parent_resolved = parent.canonicalize()?;
                Ok(parent_resolved.join(target.file_name().unwrap_or_default()))
            } else {
                Err(std::io::Error::new(
                    std::io::ErrorKind::InvalidInput,
                    "Invalid path",
                ))
            }
        })
        .map_err(|e| format!("Failed to resolve path: {}", e))?;

    // Check if the resolved path is within the working directory
    if !resolved.starts_with(&work) {
        return Err(format!(
            "Path '{}' escapes working directory '{}'",
            file_path.as_ref().display(),
            work.display()
        ));
    }

    Ok(resolved)
}

/// Check if text passes the whitelist/blacklist policy.
///
/// Whitelist is checked first - if matched, the text is allowed.
/// Then blacklist is checked - if matched, the text is denied.
///
/// Returns `(allowed, reason)`.
pub fn check_security_policy(
    text: &str,
    whitelist: &[Pattern],
    blacklist: &[Pattern],
    default_allow: bool,
) -> (bool, String) {
    // Check whitelist first - if matched, allow
    for pattern in whitelist {
        if pattern.matches(text) {
            return (true, format!("Matched whitelist pattern: {}", pattern.pattern));
        }
    }

    // Check blacklist - if matched, deny
    for pattern in blacklist {
        if pattern.matches(text) {
            return (false, format!("Matched blacklist pattern: {}", pattern.pattern));
        }
    }

    // Default action
    if default_allow {
        (true, "Default allow".to_string())
    } else {
        (false, "Default deny".to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    #[test]
    fn test_resolve_safe_path_relative() {
        let temp = TempDir::new().unwrap();
        let work = temp.path();
        let file = work.join("test.txt");
        fs::write(&file, "hello").unwrap();

        let resolved = resolve_safe_path("test.txt", work).unwrap();
        assert_eq!(resolved, file.canonicalize().unwrap());
    }

    #[test]
    fn test_resolve_safe_path_subdir() {
        let temp = TempDir::new().unwrap();
        let work = temp.path();
        let subdir = work.join("subdir");
        fs::create_dir(&subdir).unwrap();
        let file = subdir.join("test.txt");
        fs::write(&file, "hello").unwrap();

        let resolved = resolve_safe_path("subdir/test.txt", work).unwrap();
        assert_eq!(resolved, file.canonicalize().unwrap());
    }

    #[test]
    fn test_resolve_safe_path_escapes_work_dir() {
        let temp = TempDir::new().unwrap();
        let work = temp.path();

        let result = resolve_safe_path("../outside.txt", work);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("escapes working directory"));
    }

    #[test]
    fn test_resolve_safe_path_absolute_within() {
        let temp = TempDir::new().unwrap();
        let work = temp.path();
        let file = work.join("test.txt");
        fs::write(&file, "hello").unwrap();

        let resolved = resolve_safe_path(&file, work).unwrap();
        assert_eq!(resolved, file.canonicalize().unwrap());
    }

    #[test]
    fn test_resolve_safe_path_absolute_outside() {
        let temp1 = TempDir::new().unwrap();
        let temp2 = TempDir::new().unwrap();
        let work = temp1.path();
        let outside = temp2.path().join("outside.txt");
        fs::write(&outside, "hello").unwrap();

        let result = resolve_safe_path(&outside, work);
        assert!(result.is_err());
    }

    #[test]
    fn test_check_security_policy_whitelist() {
        let whitelist = vec![Pattern::wildcard("ls *")];
        let blacklist = vec![Pattern::wildcard("rm *")];

        let (allowed, reason) = check_security_policy("ls -la", &whitelist, &blacklist, true);
        assert!(allowed);
        assert!(reason.contains("whitelist"));
    }

    #[test]
    fn test_check_security_policy_blacklist() {
        let whitelist: Vec<Pattern> = vec![];
        let blacklist = vec![Pattern::wildcard("rm *")];

        let (allowed, reason) = check_security_policy("rm -rf /", &whitelist, &blacklist, true);
        assert!(!allowed);
        assert!(reason.contains("blacklist"));
    }

    #[test]
    fn test_check_security_policy_default_allow() {
        let whitelist: Vec<Pattern> = vec![];
        let blacklist: Vec<Pattern> = vec![];

        let (allowed, _) = check_security_policy("echo hello", &whitelist, &blacklist, true);
        assert!(allowed);
    }

    #[test]
    fn test_check_security_policy_default_deny() {
        let whitelist: Vec<Pattern> = vec![];
        let blacklist: Vec<Pattern> = vec![];

        let (allowed, _) = check_security_policy("echo hello", &whitelist, &blacklist, false);
        assert!(!allowed);
    }

    #[test]
    fn test_whitelist_priority_over_blacklist() {
        // If a command matches both whitelist and blacklist, whitelist wins
        let whitelist = vec![Pattern::wildcard("rm *safe*")];
        let blacklist = vec![Pattern::wildcard("rm *")];

        let (allowed, reason) = check_security_policy("rm safe_file.txt", &whitelist, &blacklist, true);
        assert!(allowed);
        assert!(reason.contains("whitelist"));
    }

    #[test]
    fn test_wildcard_pattern_matching() {
        let pattern = Pattern::wildcard("ls *");
        assert!(pattern.matches("ls -la"));
        assert!(pattern.matches("ls /tmp"));
        assert!(!pattern.matches("rm file.txt"));

        let pattern2 = Pattern::wildcard("*dangerous*");
        assert!(pattern2.matches("something_dangerous_here"));
        assert!(!pattern2.matches("safe_command"));
    }
}
